File: api/knowledge.py
"""Isolated Knowledge Base collections for the Linguistic Research API."""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING, Any, Dict, Iterable, List, Optional, Tuple

if TYPE_CHECKING:
    from api.store import Dataset, Store

DOCUMENT_COLLECTIONS = (
    "grammar",
    "expressions",
    "proverbs",
    "culture",
    "food",
    "music",
    "dance",
    "folklore",
    "stories",
    "places",
)

KNOWLEDGE_BASE = (
    "languages",
    "entries",
    *DOCUMENT_COLLECTIONS,
    "sources",
    "search",
)

LISTABLE_COLLECTIONS = ("entries", *DOCUMENT_COLLECTIONS, "sources")


def empty_knowledge() -> Dict[str, List[Dict[str, Any]]]:
    return {name: [] for name in DOCUMENT_COLLECTIONS}


def knowledge_path(json_path: Path) -> Path:
    return json_path.parent / "knowledge.json"


def load_knowledge(json_path: Path, expected_language: Optional[str]) -> Dict[str, List[Dict[str, Any]]]:
    payload = empty_knowledge()
    path = knowledge_path(json_path)
    if not path.is_file():
        return payload
    with path.open(encoding="utf-8") as handle:
        document = json.load(handle)
    for name in DOCUMENT_COLLECTIONS:
        payload[name] = _accepted_records(
            name,
            document.get(name) or [],
            expected_language,
        )
    return payload


def _accepted_records(
    collection: str,
    rows: Iterable[Dict[str, Any]],
    expected_language: Optional[str],
) -> List[Dict[str, Any]]:
    accepted: List[Dict[str, Any]] = []
    for row in rows:
        if not isinstance(row, dict):
            continue
        if not row.get("id") or not row.get("source"):
            continue
        language = row.get("language")
        if expected_language and language and language != expected_language:
            continue
        record = dict(row)
        record["collection"] = collection
        accepted.append(record)
    return accepted


def collection_path(dataset: Dataset, collection: str) -> str:
    return f"/v1/{dataset.ref.key}/{collection}"


def knowledge_base_map(origin: str) -> Dict[str, Any]:
    host = origin.rstrip("/")
    collections = []
    for name in KNOWLEDGE_BASE:
        item = {
            "id": name,
            "path": f"/v1/{name}",
            "url": f"{host}/v1/{name}",
        }
        if name == "languages":
            item["isolation"] = "Catalog of isolated lexicons. Not a merged word list."
        elif name == "search":
            item["isolation"] = "Requires dataset=. Search never crosses folders."
            item["example"] = "/v1/search?dataset=saotome/forro&q=kume"
        else:
            item["isolation"] = (
                "Content lives on one language path. "
                f"Example: /v1/saotome/forro/{name}"
            )
            item["example"] = f"/v1/saotome/forro/{name}"
        collections.append(item)
    return {
        "name": "ForroVivo Linguistic Research Knowledge Base",
        "principle": "Zero hallucination. Missing data is preferable to incorrect data.",
        "isolation": (
            "Each collection is served per isolated dataset. "
            "Parent indexes are not merged knowledge. "
            "Empty collections stay empty until a cited source fills them."
        ),
        "file": "knowledge.json next to dictionary.json in the same isolated folder",
        "collections": collections,
    }


def language_paths(dataset: Dataset) -> Dict[str, str]:
    paths = {name: collection_path(dataset, name) for name in LISTABLE_COLLECTIONS}
    paths["search"] = collection_path(dataset, "search")
    paths["lookup"] = collection_path(dataset, "lookup")
    return paths


def languages_catalog(store: Store, origin: str) -> Dict[str, Any]:
    host = origin.rstrip("/")
    languages = []
    for key in sorted(store.datasets):
        dataset = store.datasets[key]
        if dataset.kind != "lexicon":
            continue
        meta = dataset.metadata()
        languages.append(
            {
                "dataset": key,
                "language": meta.get("language"),
                "language_name": meta.get("language_name"),
                "iso_639_3": meta.get("iso_639_3"),
                "entry_count": meta.get("entry_count"),
                "knowledge": {
                    name: len(dataset.knowledge.get(name) or [])
                    for name in DOCUMENT_COLLECTIONS
                },
                "path": f"/v1/{key}",
                "url": f"{host}/v1/{key}",
                "collections": {
                    name: f"{host}{path}"
                    for name, path in language_paths(dataset).items()
                },
            }
        )
    return {
        "isolation": (
            "Each object is one lexicon. Counts are per folder. "
            "This list is not a merged dictionary."
        ),
        "languages": languages,
    }


def collection_directory(store: Store, collection: str) -> Dict[str, Any]:
    datasets = []
    for key in sorted(store.datasets):
        dataset = store.datasets[key]
        if dataset.kind != "lexicon":
            continue
        if collection == "entries":
            count = len(dataset.entries)
        elif collection == "sources":
            count = len(dataset.document.get("sources") or [])
        else:
            count = len(dataset.knowledge.get(collection) or [])
        datasets.append(
            {
                "dataset": key,
                "count": count,
                "path": collection_path(dataset, collection),
            }
        )
    return {
        "collection": collection,
        "isolation": (
            "This index lists isolated folders only. It does not merge records. "
            "Open one path to read that language."
        ),
        "datasets": datasets,
    }


def dataset_sources(dataset: Dataset) -> List[Dict[str, Any]]:
    rows = []
    for index, source in enumerate(dataset.document.get("sources") or []):
        if isinstance(source, dict):
            item = dict(source)
        else:
            item = {"source": str(source)}
        item.setdefault("id", item.get("source_file") or f"{dataset.ref.key}-source-{index + 1}")
        item["dataset"] = dataset.ref.key
        item["language"] = dataset.document.get("language")
        rows.append(item)
    return rows


def collection_records(dataset: Dataset, collection: str) -> List[Dict[str, Any]]:
    if collection == "entries":
        return list(dataset.entries)
    if collection == "sources":
        return dataset_sources(dataset)
    return list(dataset.knowledge.get(collection) or [])


def record_by_id(dataset: Dataset, collection: str, item_id: str) -> Optional[Dict[str, Any]]:
    for row in collection_records(dataset, collection):
        if row.get("id") == item_id:
            return row
    return None


def search_collection(dataset: Dataset, collection: str, query: Optional[str]) -> List[Dict[str, Any]]:
    from api.store import fold, search_entries

    rows = collection_records(dataset, collection)
    if not query:
        return rows
    if collection == "entries":
        return search_entries(dataset, query)
    needle = fold(query)
    hits = []
    for row in rows:
        blob = fold(
            " ".join(
                str(row.get(field) or "")
                for field in ("id", "title", "headword", "text", "text_pt", "text_en", "source", "source_title")
            )
        )
        if needle in blob:
            hits.append(row)
    return hits


def search_dataset(dataset: Dataset, query: str) -> List[Dict[str, Any]]:
    from api.store import fold, search_entries

    hits: List[Dict[str, Any]] = []
    for entry in search_entries(dataset, query):
        hits.append({"collection": "entries", "item": entry})
    needle = fold(query)
    if not needle:
        return hits
    for collection in DOCUMENT_COLLECTIONS:
        for row in dataset.knowledge.get(collection) or []:
            blob = fold(
                " ".join(
                    str(row.get(field) or "")
                    for field in ("id", "title", "headword", "text", "text_pt", "text_en")
                )
            )
            if needle in blob:
                hits.append({"collection": collection, "item": row})
    for source in dataset_sources(dataset):
        blob = fold(
            " ".join(
                str(source.get(field) or "")
                for field in ("id", "source", "source_title", "source_file")
            )
        )
        if needle in blob:
            hits.append({"collection": "sources", "item": source})
    return hits


def page_rows(rows: List[Dict[str, Any]], offset: int, limit: int) -> Tuple[List[Dict[str, Any]], int]:
    return rows[offset : offset + limit], len(rows)

File: api/src/knowledge.ts
import type { Dataset } from "./store";
import { fold, searchEntries } from "./store";
import { fetchGithubJson, repoHasFile } from "./github";

export const DOCUMENT_COLLECTIONS = [
  "grammar",
  "expressions",
  "proverbs",
  "culture",
  "food",
  "music",
  "dance",
  "folklore",
  "stories",
  "places",
] as const;

export type DocumentCollection = (typeof DOCUMENT_COLLECTIONS)[number];

export const KNOWLEDGE_BASE = [
  "languages",
  "entries",
  ...DOCUMENT_COLLECTIONS,
  "sources",
  "search",
] as const;

export const LISTABLE_COLLECTIONS = [
  "entries",
  ...DOCUMENT_COLLECTIONS,
  "sources",
] as const;

export type Knowledge = Record<DocumentCollection, Record<string, unknown>[]>;

export function emptyKnowledge(): Knowledge {
  return {
    grammar: [],
    expressions: [],
    proverbs: [],
    culture: [],
    food: [],
    music: [],
    dance: [],
    folklore: [],
    stories: [],
    places: [],
  };
}

function knowledgePath(jsonPath: string): string {
  const folder = jsonPath.replace(/\/dictionary\.json$/, "");
  return `${folder}/knowledge.json`;
}

function acceptedRecords(
  collection: string,
  rows: unknown,
  expectedLanguage: string | undefined,
): Record<string, unknown>[] {
  if (!Array.isArray(rows)) {
    return [];
  }
  const accepted: Record<string, unknown>[] = [];
  for (const row of rows) {
    if (!row || typeof row !== "object") {
      continue;
    }
    const record = row as Record<string, unknown>;
    if (!record.id || !record.source) {
      continue;
    }
    const language = record.language;
    if (expectedLanguage && language && language !== expectedLanguage) {
      continue;
    }
    accepted.push({ ...record, collection });
  }
  return accepted;
}

export async function loadKnowledge(
  env: Env,
  jsonPath: string,
  expectedLanguage: string | undefined,
): Promise<Knowledge> {
  const payload = emptyKnowledge();
  const path = knowledgePath(jsonPath);
  if (!(await repoHasFile(env, path))) {
    return payload;
  }
  const document = await fetchGithubJson(env, path);
  if (!document) {
    return payload;
  }
  for (const name of DOCUMENT_COLLECTIONS) {
    payload[name] = acceptedRecords(name, document[name], expectedLanguage);
  }
  return payload;
}

export function collectionPath(dataset: Dataset, collection: string): string {
  return `/v1/${dataset.ref.key}/${collection}`;
}

export function knowledgeBaseMap(origin: string): Record<string, unknown> {
  const host = origin.replace(/\/$/, "");
  const collections = KNOWLEDGE_BASE.map((name) => {
    const item: Record<string, unknown> = {
      id: name,
      path: `/v1/${name}`,
      url: `${host}/v1/${name}`,
    };
    if (name === "languages") {
      item.isolation = "Catalog of isolated lexicons. Not a merged word list.";
    } else if (name === "search") {
      item.isolation = "Requires dataset=. Search never crosses folders.";
      item.example = "/v1/search?dataset=saotome/forro&q=kume";
    } else {
      item.isolation = `Content lives on one language path. Example: /v1/saotome/forro/${name}`;
      item.example = `/v1/saotome/forro/${name}`;
    }
    return item;
  });
  return {
    name: "ForroVivo Linguistic Research Knowledge Base",
    principle: "Zero hallucination. Missing data is preferable to incorrect data.",
    isolation:
      "Each collection is served per isolated dataset. Parent indexes are not merged knowledge. Empty collections stay empty until a cited source fills them.",
    file: "knowledge.json next to dictionary.json in the same isolated folder",
    collections,
  };
}

export function languagePaths(dataset: Dataset): Record<string, string> {
  const paths: Record<string, string> = {};
  for (const name of LISTABLE_COLLECTIONS) {
    paths[name] = collectionPath(dataset, name);
  }
  paths.search = collectionPath(dataset, "search");
  paths.lookup = collectionPath(dataset, "lookup");
  return paths;
}

export function languagesCatalog(
  datasets: Dataset[],
  origin: string,
): Record<string, unknown> {
  const host = origin.replace(/\/$/, "");
  const languages = datasets
    .filter((dataset) => dataset.kind === "lexicon")
    .sort((a, b) => a.ref.key.localeCompare(b.ref.key))
    .map((dataset) => {
      const meta = dataset.metadata();
      return {
        dataset: dataset.ref.key,
        language: meta.language,
        language_name: meta.language_name,
        iso_639_3: meta.iso_639_3,
        entry_count: meta.entry_count,
        knowledge: Object.fromEntries(
          DOCUMENT_COLLECTIONS.map((name) => [
            name,
            (dataset.knowledge[name] || []).length,
          ]),
        ),
        path: `/v1/${dataset.ref.key}`,
        url: `${host}/v1/${dataset.ref.key}`,
        collections: Object.fromEntries(
          Object.entries(languagePaths(dataset)).map(([name, path]) => [
            name,
            `${host}${path}`,
          ]),
        ),
      };
    });
  return {
    isolation:
      "Each object is one lexicon. Counts are per folder. This list is not a merged dictionary.",
    languages,
  };
}

export function collectionDirectory(
  datasets: Dataset[],
  collection: string,
): Record<string, unknown> {
  const items = datasets
    .filter((dataset) => dataset.kind === "lexicon")
    .sort((a, b) => a.ref.key.localeCompare(b.ref.key))
    .map((dataset) => {
      let count = 0;
      if (collection === "entries") {
        count = dataset.entries.length;
      } else if (collection === "sources") {
        count = ((dataset.document.sources as unknown[]) || []).length;
      } else {
        count = (dataset.knowledge[collection as DocumentCollection] || []).length;
      }
      return {
        dataset: dataset.ref.key,
        count,
        path: collectionPath(dataset, collection),
      };
    });
  return {
    collection,
    isolation:
      "This index lists isolated folders only. It does not merge records. Open one path to read that language.",
    datasets: items,
  };
}

export function datasetSources(dataset: Dataset): Record<string, unknown>[] {
  const rows: Record<string, unknown>[] = [];
  const sources = (dataset.document.sources as unknown[]) || [];
  sources.forEach((source, index) => {
    const item =
      source && typeof source === "object"
        ? { ...(source as Record<string, unknown>) }
        : { source: String(source) };
    if (!item.id) {
      item.id = item.source_file || `${dataset.ref.key}-source-${index + 1}`;
    }
    item.dataset = dataset.ref.key;
    item.language = dataset.document.language;
    rows.push(item);
  });
  return rows;
}

export function collectionRecords(
  dataset: Dataset,
  collection: string,
): Record<string, unknown>[] {
  if (collection === "entries") {
    return dataset.entries;
  }
  if (collection === "sources") {
    return datasetSources(dataset);
  }
  return dataset.knowledge[collection as DocumentCollection] || [];
}

export function recordById(
  dataset: Dataset,
  collection: string,
  itemId: string,
): Record<string, unknown> | null {
  return (
    collectionRecords(dataset, collection).find((row) => row.id === itemId) ||
    null
  );
}

export function searchCollection(
  dataset: Dataset,
  collection: string,
  query: string | null,
): Record<string, unknown>[] {
  const rows = collectionRecords(dataset, collection);
  if (!query) {
    return rows;
  }
  if (collection === "entries") {
    return searchEntries(dataset, query);
  }
  const needle = fold(query);
  return rows.filter((row) => {
    const blob = fold(
      ["id", "title", "headword", "text", "text_pt", "text_en", "source", "source_title"]
        .map((field) => String(row[field] || ""))
        .join(" "),
    );
    return blob.includes(needle);
  });
}

export function searchDataset(
  dataset: Dataset,
  query: string,
): Array<{ collection: string; item: Record<string, unknown> }> {
  const hits: Array<{ collection: string; item: Record<string, unknown> }> = [];
  for (const entry of searchEntries(dataset, query)) {
    hits.push({ collection: "entries", item: entry });
  }
  const needle = fold(query);
  if (!needle) {
    return hits;
  }
  for (const collection of DOCUMENT_COLLECTIONS) {
    for (const row of dataset.knowledge[collection] || []) {
      const blob = fold(
        ["id", "title", "headword", "text", "text_pt", "text_en"]
          .map((field) => String(row[field] || ""))
          .join(" "),
      );
      if (blob.includes(needle)) {
        hits.push({ collection, item: row });
      }
    }
  }
  for (const source of datasetSources(dataset)) {
    const blob = fold(
      ["id", "source", "source_title", "source_file"]
        .map((field) => String(source[field] || ""))
        .join(" "),
    );
    if (blob.includes(needle)) {
      hits.push({ collection: "sources", item: source });
    }
  }
  return hits;
}

export function pageRows<T>(rows: T[], offset: number, limit: number): [T[], number] {
  return [rows.slice(offset, offset + limit), rows.length];
}
